### PR TITLE
Bundle select reworked

### PR DESCRIPTION
**Pull request by oskin1**

### CODE REVIEW

File: spectrum-offchain-lm/src/bundle.rs [hunk 1]
```diff
@@ -3,53 +3,40 @@ use std::sync::Arc;
...
-    async fn get_state(&self, state_id: BundleStateId) -> Option<StakingBundle>;
+    async fn get_state(&self, state_id: BundleStateId) -> Option<AsBox<IndexedStakingBundle>>;
     /// Invalidate bundle state snapshot corresponding to the given `state_id`.
     async fn invalidate(&self, state_id: BundleStateId);
+    /// Mark given bundle as permanently eliminated.
+    async fn eliminate(&self, bundle: IndexedStakingBundle);
     /// Persist confirmed state staking bundle.
-    async fn put_confirmed(&self, bundle: Confirmed<AsBox<StakingBundle>>);
-    /// Persist unconfirmed state staking bundle.
-    async fn put_unconfirmed(&self, bundle: Unconfirmed<AsBox<StakingBundle>>);
```
Comment (oskin1) [inline spectrum-offchain-lm/src/bundle.rs:31]: No need to track unconfirmed staking bundles.

File: spectrum-offchain-lm/src/bundle/rocksdb.rs [hunk 2]
```diff
@@ -1,60 +1,74 @@
...
+    prefix_bytes.extend_from_slice(&pool_id_bytes);
+    prefix_bytes.extend_from_slice(&epoch_ix_bytes);
+    prefix_bytes
+}
 
-    fn get_height_range(&self, epoch_ix: u32) -> EpochHeightRange {
-        let first = (epoch_ix - 1) * self.epoch_len;
-        let last = epoch_ix * self.epoch_len - 1;
-        EpochHeightRange { first, last }
+fn destruct_epoch_index_key(key: &[u8]) -> Option<(BundleId, u32)> {
```
Comment (kettlebell) [inline spectrum-offchain-lm/src/bundle/rocksdb.rs:35]: You mean `destructure_epoch_index_key` like `let (x:xs) = [1, 2, 3]` in Haskell right?

File: spectrum-offchain-lm/src/event_sink/handlers/bundle.rs [hunk 3]
```diff
@@ -1,82 +1,106 @@
...
+                if let Some(prog) = self.programs.lock().get(bundle.pool_id).await {
+                    let indexed_bundle = IndexedBundle::new(bundle, prog);
+                    created_bundles.insert(indexed_bundle.get_self_ref(), AsBox(bx.clone(), indexed_bundle));
                 }
-                Some(LedgerTxEvent::AppliedTx { tx, timestamp })
             }
-            ev => Some(ev),
         }
+        let consumed_keys = consumed_bundles.keys().cloned().collect::<HashSet<_>>();
+        let created_keys = consumed_bundles.keys().cloned().collect::<HashSet<_>>();
```
Comment (kettlebell) [inline spectrum-offchain-lm/src/event_sink/handlers/bundle.rs:58]:
```suggestion
        let created_keys = created_bundles.keys().cloned().collect::<HashSet<_>>();
```

**Review by kettlebell — CHANGES_REQUESTED** [review]
Looking good, just some minor changes needed.

File: spectrum-offchain-lm/src/bundle/rocksdb.rs [hunk 4]
```diff
@@ -1,60 +1,74 @@
...
 
-#[async_trait]
-impl BundleRepo for BundleRepoRocksDB {
-    fn get_epoch_length(&self) -> u32 {
-        self.epoch_len
-    }
+fn epoch_index_prefix(pool_id: PoolId, epoch_ix: u32) -> Vec<u8> {
+    let mut prefix_bytes = bincode::serialize(POOL_EPOCH_PREFIX).unwrap();
+    let pool_id_bytes = bincode::serialize(&pool_id).unwrap();
+    let neg_epoch_ix = u32::MAX - epoch_ix;
```
[the remark this answers is not in the document]
Comment (oskin1) [reply spectrum-offchain-lm/src/bundle/rocksdb.rs:28]: To maintain ascending order or indexes in rocksdb.

**Review by kettlebell — APPROVED** [review]

**Pull request merged**